Fully implement OSC 4, 10, 11, 12, 104, 110, 111, and 112

gpanders opened the pull request:
Fixes: https://github.com/mitchellh/ghostty/issues/722

Commits are broken out individually.

gpanders committed:
core: implement querying with OSC 4

gpanders committed:
termio: update foreground and background color on config change

gpanders committed:
core: implement setting colors with OSC 4, 10, and 11

mitchellh reviewed (changes requested):
Looks very good, as usual. 

There is one comment I have directly on the current implementation.

I think a broader issue I see (which doesn't have to block merge): currently if you reload config, it will override and reset the fg/bg/palette, I think? The behavior we probably want is that the config defines the _default_ values that are used if the terminal program hasn't set them. What do you think?

I'm happy to merge this first and address that later, but want to check if we should track that for the future.

gpanders committed:
core: move color parsing functions into RGB namespace

gpanders committed:
core: separate default colors from modifiable colors

Default colors are those set by the user in the config file, or an
actual default value if unset. The actual colors are modifiable and can
be changed using the OSC 4, 10, and 11 sequences.

gpanders committed:
core: implement OSC 104, 110, and 111 to reset colors

gpanders committed:
core: implement OSC 12 and OSC 112 to query/set/reset cursor color

mitchellh committed:
terminal: change mask from u256 to StaticBitSet
